### PR TITLE
Add numeric_debugger top level APIs

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130643

Summary:
Add three top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

* `prepare_for_propagation_comparison`
* `extract_results_from_loggers`
* `compare_results`

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: test/quantization/pt2e/test_numeric_debugger.py [hunk 1]
```diff
@@ -121,3 +123,40 @@ def test_re_export_preserve_handle(self):
...
+        m = TestHelperModules.Conv2dThenConv1d()
+        example_inputs = m.example_inputs()
+        m = capture_pre_autograd_graph(m, example_inputs)
+        generate_numeric_debug_handle(m)
+        m_logger = prepare_for_propagation_comparison(m)
+        ref = m(*example_inputs)
+        res = m_logger(*example_inputs)
+
+        from torch.ao.quantization.pt2e.numeric_debugger import OutputLogger
+        assert any(isinstance(m, OutputLogger) for m in m_logger.modules())
```
Comment (dulinriley) [inline test/quantization/pt2e/test_numeric_debugger.py:137]: `any` is not quite specific, perhaps we should check `all` instead? Or maybe find a node of a particular function, and make sure that a call_module after it is an OutputLogger?

File: test/quantization/pt2e/test_numeric_debugger.py [hunk 2]
```diff
@@ -121,3 +123,40 @@ def test_re_export_preserve_handle(self):
...
+
+        self.assertEqual(res, ref)
```
Comment (dulinriley) [inline test/quantization/pt2e/test_numeric_debugger.py:139]: You should use torch.testing.assert_close instead

File: test/quantization/pt2e/test_numeric_debugger.py [hunk 3]
```diff
@@ -121,3 +123,40 @@ def test_re_export_preserve_handle(self):
...
+        m = TestHelperModules.Conv2dThenConv1d()
+        example_inputs = m.example_inputs()
+        m = capture_pre_autograd_graph(m, example_inputs)
+        generate_numeric_debug_handle(m)
+        m_ref_logger = prepare_for_propagation_comparison(m)
+
+        quantizer = XNNPACKQuantizer().set_global(
+            get_symmetric_quantization_config(is_per_channel=False)
+        )
+        m = prepare_pt2e(m, quantizer)
```
Comment (dulinriley) [inline test/quantization/pt2e/test_numeric_debugger.py:151]: You might want to have some kind of check/assert in prepare_pt2e to ensure the loggers *aren't* in there.
The loggers will block both dead code elimination, pattern matching, and optimizations that remove nodes

File: test/quantization/pt2e/test_numeric_debugger.py [hunk 4]
```diff
@@ -121,3 +123,40 @@ def test_re_export_preserve_handle(self):
...
+        m(*example_inputs)
+        m = convert_pt2e(m)
+        debug_handle_map = _extract_debug_handles(m)
```
Comment (dulinriley) [inline test/quantization/pt2e/test_numeric_debugger.py:154]: Why is this API private (leading underscore)?

File: test/quantization/pt2e/test_numeric_debugger.py [hunk 5]
```diff
@@ -121,3 +123,40 @@ def test_re_export_preserve_handle(self):
...
+        m_quant_logger = prepare_for_propagation_comparison(m)
+
+        m_ref_logger(*example_inputs)
+        m_quant_logger(*example_inputs)
+        results = extract_results_from_loggers(m_ref_logger, m_quant_logger)
+        for node, node_summary in results.items():
+            if len(node_summary.results) > 0:
+                self.assertTrue(node_summary.results[0].sqnr >= 35)
```
Comment (dulinriley) [inline test/quantization/pt2e/test_numeric_debugger.py:162]:
```suggestion
                self.assertGreaterEqual(node_summary.results[0].sqnr, 35)
```

File: torch/ao/quantization/pt2e/numeric_debugger.py [hunk 6]
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
         if node.op != "placeholder" and NUMERIC_DEBUG_HANDLE_KEY not in node.meta:
             node.meta[NUMERIC_DEBUG_HANDLE_KEY] = unique_id
             unique_id += 1
+
+
+class OutputLogger(torch.nn.Module):
+    """
+    Base class for capturing output values.
+    """
```
Comment (dulinriley) [inline torch/ao/quantization/pt2e/numeric_debugger.py:28]: Can you add a better description? What types of output values does it capture? When would it not work?
I'd recommend focusing on the cases specific to prepare/convert for now

File: torch/ao/quantization/pt2e/numeric_debugger.py [hunk 7]
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
         if node.op != "placeholder" and NUMERIC_DEBUG_HANDLE_KEY not in node.meta:
             node.meta[NUMERIC_DEBUG_HANDLE_KEY] = unique_id
             unique_id += 1
+
+
+class OutputLogger(torch.nn.Module):
```
Comment (dulinriley) [inline torch/ao/quantization/pt2e/numeric_debugger.py:25]: Does this need a `backwards_compatibility` annotation? Is it a public torch API?

File: torch/ao/quantization/pt2e/numeric_debugger.py [hunk 8]
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
...
+    Base class for capturing output values.
+    """
+
+    # Mark as impure so that calls to it will not be removed during DCE.
+    _is_impure = True
+
+    def __init__(
+        self,
+        node_name: str,
+        nn_module_stack: object,
```
Comment (dulinriley) [inline torch/ao/quantization/pt2e/numeric_debugger.py:36]: You could make this default to None, in case other users don't have a stack available. It's an optional part of the logger, just to make identifying stuff easier. Same with `node_name` too, only `debug_handle` is required

File: torch/ao/quantization/pt2e/numeric_debugger.py [hunk 9]
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
...
+        return x
+
+    def __repr__(self) -> str:
+        clean_dict = {
+            k: v
+            for k, v in self.__dict__.items()
+            # skip torch.nn.Module keys
+            if (k != "training") and not k.startswith("_")
+        }
+        return f"OutputLogger({clean_dict})"
```
Comment (dulinriley) [inline torch/ao/quantization/pt2e/numeric_debugger.py:57]: Why include all this stuff?
Also, I think typically subclasses of Module use "extra_repr" instead
```suggestion
    def __repr__(self) -> str:
        return f"OutputLogger(node_name={self.node_name}, nn_module_stack={self.nn_module_stack}, debug_handle={self.debug_handle}, num_stats={len(self.stats)})"
```

File: torch/ao/quantization/pt2e/numeric_debugger.py [hunk 10]
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
...
+
+
+def _insert_logger(model: GraphModule, node: Node, debug_handle: str) -> None:
```
Comment (dulinriley) [inline torch/ao/quantization/pt2e/numeric_debugger.py:60]: nit: might want to return the OutputLogger instance, or the logger node. Some callers might want to access it after-the-fact

File: torch/ao/quantization/pt2e/numeric_debugger.py [hunk 11]
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
...
+    The OutputLogger will contain the debug_handle which can be used to compare
+    graphs after transforms"""
+
+    # to avoid circular dep
+    from torch.ao.quantization.fx.utils import get_new_attr_name_with_prefix
+
+    # add a logger after the node
+    with model.graph.inserting_after(node):
+        get_new_attr_name = get_new_attr_name_with_prefix(f"{node.name}_logger")
+        logger_name = get_new_attr_name(model)
```
Comment (dulinriley) [inline torch/ao/quantization/pt2e/numeric_debugger.py:72]: Maybe assert that this attribute name doesn't exist yet?

File: torch/ao/quantization/pt2e/numeric_debugger.py [hunk 12]
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
...
+        _insert_logger(model, n, numeric_debug_handle)
+
+    model.recompile()
+    return model
+
+@dataclass(frozen=True)
+class QuantizationComparisonResult:
+    actual: torch.Tensor
+    ref: torch.Tensor
+
```
Comment (dulinriley) [inline torch/ao/quantization/pt2e/numeric_debugger.py:103]: I recommend adding `__post_init__` and assert that the actual/ref are really tensors

File: torch/ao/quantization/pt2e/numeric_debugger.py [hunk 13]
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
...
+        )
+
+
+@dataclass(frozen=True)
+class NodeAccuracySummary:
+    handle: str
+    actual_node_name: str
+    actual_module_stack: str
+    ref_node_name: str
+    ref_module_stack: str
```
Comment (dulinriley) [inline torch/ao/quantization/pt2e/numeric_debugger.py:133]: nit: swap this order.

Also, for the expansion of this concept to "one ref node, possibly multiple comparison nodes", you can have a list of actual_name and actual_stack.

But that can be a follow-up PR if you want
```suggestion
    ref_node_name: str
    ref_module_stack: str
    actual_node_name: str
    actual_module_stack: str
```

File: torch/ao/quantization/pt2e/numeric_debugger.py [hunk 14]
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
...
+        """Returns a very detailed string for the results including all tensors.
+        If you don't want to see this, just use str() or repr() on this object instead
+        to get a more concise summary."""
+        s = f"{repr(self)}\ndetails:\n"
+        for q in self.results:
+            s += q.details() + "\n"
+        return s
+
+
+def _module_stack_to_str(module_stack: object) -> str:
```
Comment (dulinriley) [inline torch/ao/quantization/pt2e/numeric_debugger.py:146]: You might want to put this function somewhere else, I'm sure other people like to visualize the nn_module_stack

File: torch/ao/quantization/pt2e/numeric_debugger.py [hunk 15]
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
...
+
+    Returns:
+        A dict is keyed by the node name from the reference (first) model given,
+        and the values are a summary of all the tensors that passed through each
+    logger and statistics about how different they are from each other."""
+    # Results maps debug handle to a tensor list for each model being compared.
+    ref_handles: Dict[str, Tuple[str, object, List[torch.Tensor]]] = {}
+    actual_handles: Dict[str, Tuple[str, object, List[torch.Tensor]]] = {}
+    for _name, module in ref_model.named_children():
+        if isinstance(module, OutputLogger):
```
Comment (dulinriley) [inline torch/ao/quantization/pt2e/numeric_debugger.py:172]: I recommend checking if `len(module.stats) > 0` here, to avoid comparing empty loggers. Maybe a warning can be emitted in that case, the user didn't run `forward`

**Review by dulinriley — APPROVED** [review]

**Review by tarun292 — APPROVED** [review]

File: test/quantization/pt2e/test_numeric_debugger.py (hunk 1 above)
Comment (jerryzh168) [reply test/quantization/pt2e/test_numeric_debugger.py:137]: sure, we can't check all because not all modules are output loggers, but I can check the node name of loggers contains conv2d

File: test/quantization/pt2e/test_numeric_debugger.py (hunk 2 above)
Comment (jerryzh168) [reply test/quantization/pt2e/test_numeric_debugger.py:139]: why? adding logger is not going to change the result right

File: test/quantization/pt2e/test_numeric_debugger.py (hunk 3 above)
Comment (jerryzh168) [reply test/quantization/pt2e/test_numeric_debugger.py:151]: prepare_pt2e should not be aware of `OutputLogger` I think, this has to be enforced by user I feel

File: test/quantization/pt2e/test_numeric_debugger.py (hunk 4 above)
Comment (jerryzh168) [reply test/quantization/pt2e/test_numeric_debugger.py:154]: this is a test-only helper function, not user facing API

File: torch/ao/quantization/pt2e/numeric_debugger.py (hunk 7 above)
Comment (jerryzh168) [reply torch/ao/quantization/pt2e/numeric_debugger.py:25]: this is a public API, I'll add it to `torch.ao.quantization`, but we have not been using `backwards_compatibility` decorators.

File: torch/ao/quantization/pt2e/numeric_debugger.py (hunk 13 above)
Comment (jerryzh168) [reply torch/ao/quantization/pt2e/numeric_debugger.py:133]: I feel we can design this part when there is a real use case

File: torch/ao/quantization/pt2e/numeric_debugger.py (hunk 14 above)
Comment (jerryzh168) [reply torch/ao/quantization/pt2e/numeric_debugger.py:146]: this seems to be a small util, we can move when other people also needs this I think

File: test/quantization/pt2e/test_numeric_debugger.py (hunk 4 above)
Comment (jerryzh168) [reply test/quantization/pt2e/test_numeric_debugger.py:155]: sorry, this can be removed now actually
